examples/psp22-extension/runtime/psp22-extension-example.rs

```diff
@@ -0,0 +1,414 @@
+        .saturating_add(<T as pallet_assets::Config>::WeightInfo::approve_transfer());
+    // debug_message weight is a good approximation of the additional overhead of going from
+    // contract layer to substrate layer.
+    let overhead = Weight::from_ref_time(
+        <T as pallet_contracts::Config>::Schedule::get()
+            .host_fn_weights
+            .debug_message,
+    );
+    let charged_weight = env.charge_weight(base_weight.saturating_add(overhead))?;
+    trace!(
+        target: "runtime",
+        "[ChainExtension]|call|decrease_allowance / charge_weight:{:?}",
+        charged_weight
+    );
+
+    let owner = env.ext().caller();
+    let mut allowance =
+        <pallet_assets::Pallet<T> as AllowanceInspect<T::AccountId>>::allowance(
+            input.asset_id,
+            owner,
+            &input.spender,
+        );
+    <pallet_assets::Pallet<T>>::cancel_approval(
+        RawOrigin::Signed(owner.clone()).into(),
+        input.asset_id,
+        T::Lookup::unlookup(input.spender.clone()),
+    )
+    .map_err(convert_err(
+        "ChainExtension failed to call decrease_allowance",
+    ))?;
+    allowance.saturating_reduce(input.value);
+    if allowance.is_zero() {
+        // If reduce value was less or equal than existing allowance, it should stay none.
+        env.adjust_weight(
+            charged_weight,
+            <T as pallet_assets::Config>::WeightInfo::cancel_approval()
+                .saturating_add(overhead),
+        );
+        return Ok(());
+    }
+    <pallet_assets::Pallet<T> as AllowanceMutate<T::AccountId>>::approve(
+        input.asset_id,
+        owner,
+        &input.spender,
+        allowance,
+    )
+    .map_err(convert_err(
+        "ChainExtension failed to call decrease_allowance",
+    ))?;
+
+    trace!(
+        target: "runtime",
+        "[ChainExtension]|call|decrease_allowance"
+    );
+
+    Ok(())
+}
+
+impl<T> ChainExtension<T> for Psp22Extension
+where
+    T: pallet_assets::Config + pallet_contracts::Config,
+    <T as SysConfig>::AccountId: UncheckedFrom<<T as SysConfig>::Hash> + AsRef<[u8]>,
+{
+    fn call<E: Ext>(
+        &mut self,
+        env: Environment<E, InitState>,
+    ) -> Result<RetVal, DispatchError>
+    where
+        E: Ext<T = T>,
+        <E::T as SysConfig>::AccountId:
+            UncheckedFrom<<E::T as SysConfig>::Hash> + AsRef<[u8]>,
+    {
+        let func_id = FuncId::try_from(env.func_id())?;
+        match func_id {
+            FuncId::Metadata(func_id) => metadata::<T, E>(func_id, env)?,
+            FuncId::Query(func_id) => query::<T, E>(func_id, env)?,
+            FuncId::Transfer => transfer::<T, E>(env)?,
+            FuncId::TransferFrom => transfer_from::<T, E>(env)?,
+            // FIXME: This is a bit of a shortcut. It was made because the documentation
+            //        for Mutate::approve does not specify the result of subsequent calls.
```

Comment:
Please don't leave `FIXME`s (or `TODO`s) in the code. File an issue and we can address it in a follow up PR

Reply:
OK, I'll open an issue as soon as this PR is merged.